lib/spack/spack/detection/path.py

```diff
@@ -52,13 +54,143 @@ def executables_in_path(path_hints=None):
     return path_to_exe
 
 
+def libraries_in_ld_library_path(path_hints=None):
+    """Get the paths of all libraries available from LD_LIBRARY_PATH.
+
+    For convenience, this is constructed as a dictionary where the keys are
+    the library paths and the values are the names of the libraries
+    (i.e. the basename of the library path).
+
+    There may be multiple paths with the same basename. In this case it is
+    assumed there are two different instances of the library.
+
+    Args:
+        path_hints (list): list of paths to be searched. If None the list will be
+            constructed based on the LD_LIBRARY_PATH environment variable.
+    """
+    path_hints = path_hints or spack.util.environment.get_path('LD_LIBRARY_PATH')
+    search_paths = llnl.util.filesystem.search_paths_for_libraries(*path_hints)
+
+    path_to_lib = {}
+    # Reverse order of search directories so that a lib in the first
+    # LD_LIBRARY_PATH entry overrides later entries
+    for search_path in reversed(search_paths):
+        for lib in os.listdir(search_path):
+            lib_path = os.path.join(search_path, lib)
+            if is_readable(lib_path):
+                path_to_lib[lib_path] = lib
+    return path_to_lib
+
+
 def _group_by_prefix(paths):
     groups = collections.defaultdict(set)
     for p in paths:
         groups[os.path.dirname(p)].add(p)
     return groups.items()
 
 
+def by_library(packages_to_check, path_hints=None):
```

Comment:
Can you add a TODO comment here to consolidate this with `by_executable`? Namely:

* Packages should be able to define both .libraries and .executables in the future
* `determine_spec_details` should get all relevant libraries and executables in one call